### PR TITLE
fix: fixed select empty folders in select mode (Issue #1864)

### PR DESCRIPTION
**Pull request by denys-kolomiitsev**
**Description:**

- Fixed selection of the empty folders is select mode in Chatbar and Promptbar.
Issues:

- Issue #1864 

**UI changes**

image

**Checklist:**

- [x] the pull request name complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] the pull request name starts with `fix(<scope>):`, `feat(<scope>):`, `feature(<scope>):`, `chore(<scope>):`, `hotfix(<scope>):` or `e2e(<scope>):`. If contains breaking changes then the pull request name must start with `fix(<scope>)!:`, `feat(<scope>)!:`, `feature(<scope>)!:`, `chore(<scope>)!:`, `hotfix(<scope>)!:` or `e2e(<scope>)!:` where `<scope>` is name of affected project: `chat`, `chat-e2e`, `overlay`, `shared`, `sandbox-overlay`, etc.
- [x] the pull request name ends with `(Issue #<TICKET_ID>)` (comma-separated list of issues)
- [x] I confirm that do not share any confidential information like API keys or any other secrets and private URLs

### CODE REVIEW

**Comment by denys-kolomiitsev** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-chat-pr-1907.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by taa-epam** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-chat-pr-1907.nightly-test.deltixhub.io
>E2E tests status: success

**Comment by taa-epam** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-chat-pr-1907.nightly-test.deltixhub.io
>E2E tests status: success

**Comment by taa-epam** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-chat-pr-1907.nightly-test.deltixhub.io
>E2E tests status: success

**Comment by taa-epam** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-chat-pr-1907.nightly-test.deltixhub.io
>E2E tests status: success

File: apps/chat/src/store/conversations/conversations.reducers.ts [hunk 1]
```diff
@@ -798,7 +800,36 @@ export const conversationsSlice = createSlice({
...
       }
     },
+    setAllChosenEmptyFolders: (state) => {
+      if (state.searchTerm) {
+        return state;
+      }
+      state.chosenEmptyFoldersIds = state.folders
+        .filter(
+          (folder) =>
+            !isEntityOrParentsExternal(state, folder, FeatureType.Chat) &&
```
Comment (Alexander-Kezik) [inline apps/chat/src/store/conversations/conversations.reducers.ts:809]: By the way...
Is it possible to delete shared folders? Maybe we need to discuss this?

**Review by Alexander-Kezik — DISMISSED** [review]

**Comment by taa-epam** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-chat-pr-1907.nightly-test.deltixhub.io
>E2E tests status: success

File: apps/chat/src/store/conversations/conversations.reducers.ts (hunk 1 above)
Comment (denys-kolomiitsev) [reply apps/chat/src/store/conversations/conversations.reducers.ts:809]: It is possible to delete shared folders as well as conversations.
We need to discuss this.

Comment (IlyaBondar) [reply apps/chat/src/store/conversations/conversations.reducers.ts:809]: we don't allow to multi-select Shared folders at this moment. It was decided on daily meeting

**Comment by taa-epam** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-chat-pr-1907.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by taa-epam** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-chat-pr-1907.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by taa-epam** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-chat-pr-1907.nightly-test.deltixhub.io
>E2E tests status: failed

File: apps/chat/src/components/Chatbar/ChatFolders.tsx [hunk 2]
```diff
@@ -110,6 +110,14 @@ const ChatFolderTemplate = ({
     ConversationsSelectors.selectSelectedItems,
   );
 
+  const emptyFoldersIds = useAppSelector(
+    ConversationsSelectors.selectEmptyFolderIds,
+  );
+
+  const isFolderEmpty = useAppSelector((state) =>
+    ConversationsSelectors.selectIsFolderEmpty(state, folder.id),
+  );
```
Comment (IlyaBondar) [inline apps/chat/src/components/Chatbar/ChatFolders.tsx:119]: Why not `const isFolderEmpty = emptyFoldersIds.includes(folder.id);`?

File: apps/chat/src/components/Chatbar/ChatbarSettings.tsx [hunk 3]
```diff
@@ -99,6 +99,7 @@ export const ChatbarSettings = () => {
         Icon: IconSquareCheck,
         onClick: () => {
           dispatch(ConversationsActions.setAllChosenConversations());
+          dispatch(ConversationsActions.setAllChosenEmptyFolders());
```
Comment (IlyaBondar) [inline apps/chat/src/components/Chatbar/ChatbarSettings.tsx]: Why do we need separate action for that? We can do it into `setAllChosenConversations`

File: apps/chat/src/components/Chatbar/ChatbarSettings.tsx [hunk 4]
```diff
@@ -108,6 +109,7 @@ export const ChatbarSettings = () => {
         Icon: IconSquareOff,
         onClick: () => {
           dispatch(ConversationsActions.resetChosenConversations());
+          dispatch(ConversationsActions.resetChosenEmptyFolders());
```
Comment (IlyaBondar) [inline apps/chat/src/components/Chatbar/ChatbarSettings.tsx]: Why do we need separate action for that? We can do it into `resetChosenConversations`

File: apps/chat/src/components/Promptbar/components/PromptbarSettings.tsx [hunk 5]
```diff
@@ -50,6 +50,7 @@ export function PromptbarSettings() {
         Icon: IconSquareCheck,
         onClick: () => {
           dispatch(PromptsActions.setAllChosenPrompts());
+          dispatch(PromptsActions.setAllChosenEmptyFolders());
```
Comment (IlyaBondar) [inline apps/chat/src/components/Promptbar/components/PromptbarSettings.tsx]: Why do we need separate action for that? We can do it into `setAllChosenPrompts`

File: apps/chat/src/components/Promptbar/components/PromptbarSettings.tsx [hunk 6]
```diff
@@ -58,6 +59,7 @@ export function PromptbarSettings() {
         Icon: IconSquareOff,
         onClick: () => {
           dispatch(PromptsActions.resetChosenPrompts());
+          dispatch(PromptsActions.resetChosenEmptyFolders());
```
Comment (IlyaBondar) [inline apps/chat/src/components/Promptbar/components/PromptbarSettings.tsx]: Why do we need separate action for that? We can do it into `resetChosenPrompts`

File: apps/chat/src/components/Chatbar/ChatFolders.tsx (hunk 2 above)
Comment (IlyaBondar) [reply apps/chat/src/components/Chatbar/ChatFolders.tsx:119]: probably with `useMemo`

File: apps/chat/src/components/Promptbar/components/PromptFolders.tsx [hunk 7]
```diff
@@ -97,6 +97,12 @@ const PromptFolderTemplate = ({
     (state) => PromptsSelectors.selectChosenFolderIds(state, prompts),
   );
 
+  const emptyFoldersIds = useAppSelector(PromptsSelectors.selectEmptyFolderIds);
+
+  const isFolderEmpty = useAppSelector((state) =>
+    PromptsSelectors.selectIsFolderEmpty(state, folder.id),
+  );
```
Comment (IlyaBondar) [inline apps/chat/src/components/Promptbar/components/PromptFolders.tsx:104]: Why not `const isFolderEmpty = emptyFoldersIds.includes(folder.id);` (with `useMemo`)?

File: apps/chat/src/store/conversations/conversations.epics.ts [hunk 8]
```diff
@@ -2733,11 +2737,13 @@ const deleteChosenConversationsEpic: AppEpic = (action$, state$) =>
...
               (folder) =>
                 !fullyChosenFolderIds.includes(`${folder.id}/`) &&
-                conversations.some((c) => c.id.startsWith(`${folder.id}/`)),
+                (conversations.some((c) => c.id.startsWith(`${folder.id}/`)) ||
+                  emptyFoldersIds.some((id) => id === folder.id)),
             ),
           }),
         ),
         of(ConversationsActions.resetChosenConversations()),
+        of(ConversationsActions.resetChosenEmptyFolders()),
```
Comment (IlyaBondar) [inline apps/chat/src/store/conversations/conversations.epics.ts]: Why do we need separate action for that? We can do it into `resetChosenConversations`

File: apps/chat/src/store/prompts/prompts.epics.ts [hunk 9]
```diff
@@ -785,11 +788,13 @@ const deleteChosenPromptsEpic: AppEpic = (action$, state$) =>
...
               (folder) =>
                 !fullyChosenFolderIds.includes(`${folder.id}/`) &&
-                prompts.some((p) => p.id.startsWith(`${folder.id}/`)),
+                (prompts.some((p) => p.id.startsWith(`${folder.id}/`)) ||
+                  emptyFoldersIds.some((id) => id === folder.id)),
             ),
           }),
         ),
         of(PromptsActions.resetChosenPrompts()),
+        of(PromptsActions.resetChosenEmptyFolders()),
```
Comment (IlyaBondar) [inline apps/chat/src/store/prompts/prompts.epics.ts]: Why do we need separate action for that? We can do it into `resetChosenPrompts`

File: apps/chat/src/components/Chatbar/ChatFolders.tsx (hunk 2 above)
Comment (denys-kolomiitsev) [reply apps/chat/src/components/Chatbar/ChatFolders.tsx:119]: I think it is better to have separate selector for this for the reusability.

File: apps/chat/src/components/Chatbar/ChatbarSettings.tsx (hunk 3 above)
Comment (denys-kolomiitsev) [reply apps/chat/src/components/Chatbar/ChatbarSettings.tsx:102]: I think it's better to implement separate actions for improved separation of concerns, testability, and reusability.

File: apps/chat/src/components/Chatbar/ChatbarSettings.tsx (hunk 4 above)
Comment (denys-kolomiitsev) [reply apps/chat/src/components/Chatbar/ChatbarSettings.tsx:112]: I think it's better to implement separate actions for improved separation of concerns, testability, and reusability.

File: apps/chat/src/components/Promptbar/components/PromptbarSettings.tsx (hunk 5 above)
Comment (denys-kolomiitsev) [reply apps/chat/src/components/Promptbar/components/PromptbarSettings.tsx:53]: I think it's better to implement separate actions for improved separation of concerns, testability, and reusability.

File: apps/chat/src/components/Promptbar/components/PromptbarSettings.tsx (hunk 6 above)
Comment (denys-kolomiitsev) [reply apps/chat/src/components/Promptbar/components/PromptbarSettings.tsx:62]: I think it's better to implement separate actions for improved separation of concerns, testability, and reusability.

File: apps/chat/src/store/conversations/conversations.epics.ts (hunk 8 above)
Comment (denys-kolomiitsev) [reply apps/chat/src/store/conversations/conversations.epics.ts:2746]: I think it's better to implement separate actions for improved separation of concerns, testability, and reusability.

File: apps/chat/src/store/prompts/prompts.epics.ts (hunk 9 above)
Comment (denys-kolomiitsev) [reply apps/chat/src/store/prompts/prompts.epics.ts:797]: I think it's better to implement separate actions for improved separation of concerns, testability, and reusability.

File: apps/chat/src/components/Promptbar/components/PromptFolders.tsx (hunk 7 above)
Comment (denys-kolomiitsev) [reply apps/chat/src/components/Promptbar/components/PromptFolders.tsx:104]: I think it is better to have separate selector for this for the reusability.

File: apps/chat/src/store/conversations/conversations.epics.ts (hunk 8 above)
Comment (IlyaBondar) [reply apps/chat/src/store/conversations/conversations.epics.ts]: These actions aren't used separately and it's bad practice to dispatch a lot of actions into the same reducer instead do all stuff on one action. This action means "reset all selections". Why do we need to dospatch two of them in each places?

**Comment by taa-epam** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-chat-pr-1907.nightly-test.deltixhub.io
>E2E tests status: success

File: apps/chat/src/components/Chatbar/ChatbarSettings.tsx (hunk 3 above)
Comment (denys-kolomiitsev) [reply apps/chat/src/components/Chatbar/ChatbarSettings.tsx]: Code changed as suggested.

File: apps/chat/src/components/Chatbar/ChatbarSettings.tsx (hunk 4 above)
Comment (denys-kolomiitsev) [reply apps/chat/src/components/Chatbar/ChatbarSettings.tsx]: Code chsnged as suggested.

File: apps/chat/src/components/Promptbar/components/PromptbarSettings.tsx (hunk 5 above)
Comment (denys-kolomiitsev) [reply apps/chat/src/components/Promptbar/components/PromptbarSettings.tsx]: Code changed as suggested.

File: apps/chat/src/components/Promptbar/components/PromptbarSettings.tsx (hunk 6 above)
Comment (denys-kolomiitsev) [reply apps/chat/src/components/Promptbar/components/PromptbarSettings.tsx]: Code changed as suggested.

File: apps/chat/src/store/conversations/conversations.epics.ts (hunk 8 above)
Comment (denys-kolomiitsev) [reply apps/chat/src/store/conversations/conversations.epics.ts]: Code changed as suggested.
Thank you for the explanation!

File: apps/chat/src/store/prompts/prompts.epics.ts (hunk 9 above)
Comment (denys-kolomiitsev) [reply apps/chat/src/store/prompts/prompts.epics.ts]: Code changed as suggested.

**Comment by IlyaBondar** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-chat-pr-1907.nightly-test.deltixhub.io
>E2E tests status: success

**Review by IlyaBondar — APPROVED** [review]

**Pull request merged**